src/library/scala/collection/immutable/Queue.scala

```diff
@@ -108,8 +108,14 @@ sealed class Queue[+A] protected(protected val in: List[A], protected val out: L
 
   override def ++[B >: A, That](that: GenTraversableOnce[B])(implicit bf: CanBuildFrom[Queue[A], B, That]): That = {
     if (bf eq Queue.ReusableCBF) {
-      val thatQueue = that.asInstanceOf[Queue[B]]
-      new Queue[B](thatQueue.in ++ (thatQueue.out reverse_::: this.in), this.out).asInstanceOf[That]
+      val newIn =
+        if (that.isInstanceOf[Queue[_]]) {
+          val thatQueue: Queue[B] = that.asInstanceOf[Queue[B]]
+          thatQueue.in ++ (thatQueue.out reverse_::: this.in)
+        } else {
+          (new ListBuffer[B] ++= that.seq).prependToList(this.in)
```

Comment:
@mingthemerciless - You can try running collections-laws for a start, which ought to catch these sorts of things, and says how good its coverage is.  If it doesn't catch stuff like this, we should add a law.  We also should include it in the automatic test suite before release, instead of having me intermittently running it manually.  But I'm making it more friendly for other contributors (by making more of it Scala code instead of doing weird text replacement off of files with unique syntax), so maybe not just yet.

One of the problems with the test cases for the rest of the compiler and library is that there is no good way to get a picture of the test coverage.  The policy is generally to write sufficient tests to cover the interestingly different cases when fixing something (so at least things always get better), but that isn't clearly stated or carefully-enough checked for.  Getting an extra pair of eyes on PRs before they're merged is really helpful in this regard, because it's all too easy to glance at something quickly and think, "that looks okay".  (To help with that, just look for open PRs.)